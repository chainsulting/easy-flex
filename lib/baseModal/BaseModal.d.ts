import { FC, HTMLAttributes } from 'react';
import { AbsoluteSize, Color } from '../types';
export interface BaseModalProps extends HTMLAttributes<HTMLDivElement> {
    backgroundColor?: Color;
    /** Sets blur for the content covered by the modal background. */
    blur?: boolean | AbsoluteSize;
    blurElementId?: string;
    containerElementId?: string;
    /** Called if the modal background is clicked. */
    onClose: () => void;
}
export declare const BaseModal: FC<BaseModalProps>;
