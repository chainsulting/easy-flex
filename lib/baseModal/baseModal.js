"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseModal = void 0;
var jsx_runtime_1 = require("react/jsx-runtime");
var react_1 = require("react");
var react_dom_1 = require("react-dom");
var styled_components_1 = __importDefault(require("styled-components"));
var utils_1 = require("../utils");
var Background = styled_components_1.default.div(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n\tdisplay: flex;\n\tbox-sizing: border-box;\n\tflex-direction: row;\n\tposition: fixed;\n\ttop: 0;\n\tleft: 0;\n\tright: 0;\n\tbottom: 0;\n\tmin-height: 100vh;\n\tmin-width: 100vw;\n\tbackground-color: ", ";\n\talign-items: center;\n\tjustify-content: center;\n\tmargin: 0;\n\tpadding: 0;\n"], ["\n\tdisplay: flex;\n\tbox-sizing: border-box;\n\tflex-direction: row;\n\tposition: fixed;\n\ttop: 0;\n\tleft: 0;\n\tright: 0;\n\tbottom: 0;\n\tmin-height: 100vh;\n\tmin-width: 100vw;\n\tbackground-color: ", ";\n\talign-items: center;\n\tjustify-content: center;\n\tmargin: 0;\n\tpadding: 0;\n"])), function (_a) {
    var backgroundColor = _a["data-background-color"];
    return backgroundColor;
});
var BaseModal = function (_a) {
    var children = _a.children, blur = _a.blur, onClose = _a.onClose, props = __rest(_a, ["children", "blur", "onClose"]);
    var theme = (0, utils_1.useEasyFlexTheme)();
    var backgroundElement = (0, react_1.useRef)(null);
    var handleClick = (0, react_1.useCallback)(function (event) {
        if (event.target === backgroundElement.current) {
            onClose();
        }
    }, [onClose]);
    (0, react_1.useEffect)(function () {
        if (typeof blur === 'number' || (blur !== false && theme.modal.blur)) {
            var styleElement_1 = document.createElement('style');
            styleElement_1.textContent = "\n\t\t\t\t#".concat(theme.modal.blurElementId, " {\n\t\t\t\t\tfilter: blur(").concat(typeof blur === 'number' ? (0, utils_1.toPx)(blur) : (0, utils_1.toPx)(theme.modal.blur), ");\n\t\t\t\t}\n\t\t\t");
            document.head.append(styleElement_1);
            return function () {
                document.head.removeChild(styleElement_1);
            };
        }
    }, [blur, theme]);
    var container = (0, react_1.useMemo)(function () { var _a; return (_a = document.getElementById(theme.modal.containerElementId)) !== null && _a !== void 0 ? _a : document.body; }, [theme]);
    return (0, react_dom_1.createPortal)((0, jsx_runtime_1.jsx)(Background, __assign({ ref: backgroundElement, "data-background-color": theme.modal.backgroundColor, onClick: handleClick }, props, { children: children })), container);
};
exports.BaseModal = BaseModal;
var templateObject_1;
