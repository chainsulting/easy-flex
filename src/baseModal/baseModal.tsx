import React, { FC, HTMLAttributes, MouseEvent, useCallback, useEffect, useMemo, useRef } from 'react';
import { createPortal } from 'react-dom';
import styled from 'styled-components';
import { toPx, useEasyFlexTheme } from '../utils';

const Background = styled.div<{
	'data-background-color': string;
	'data-max-child-height'?: string;
	'data-max-child-width'?: string;
}>`
	display: flex;
	box-sizing: border-box;
	position: fixed;
	top: 0;
	left: 0;
	right: 0;
	bottom: 0;
	min-height: 100vh;
	min-width: 100vw;
	background-color: ${({ 'data-background-color': backgroundColor }) => backgroundColor};
	align-items: center;
	justify-content: center;
	margin: 0;
	padding: 0;

	& :only-child {
		max-height: ${({ 'data-max-child-height': maxChildHeight }) => maxChildHeight};
		max-width: ${({ 'data-max-child-width': maxChildWidth }) => maxChildWidth};
	}
`;

export interface IBaseModalProps extends HTMLAttributes<HTMLDivElement> {
	/** If true, the size of the content is limited to the viewport's size. */
	keepChildInside?: boolean;
	/** Sets blur for the content covered by the modal background. */
	blur?: boolean | number;
	/** Called if the modal background is clicked. */
	onClose: () => void;
}

export const BaseModal: FC<IBaseModalProps> = ({ children, keepChildInside = false, blur, onClose, ...props }) => {
	const theme = useEasyFlexTheme();

	const backgroundElement = useRef<HTMLDivElement>(null);

	const handleClick = useCallback(
		(event: MouseEvent<HTMLDivElement>) => {
			if (event.target === backgroundElement.current) {
				onClose();
			}
		},
		[onClose]
	);

	const maxChildHeight = useMemo<'100vh' | undefined>(() => (keepChildInside ? '100vh' : undefined), [keepChildInside]);

	const maxChildWidth = useMemo<'100vw' | undefined>(() => (keepChildInside ? '100vw' : undefined), [keepChildInside]);

	useEffect(() => {
		if (typeof blur === 'number' || (blur !== false && theme.modal.blur)) {
			const styleElement = document.createElement('style');
			styleElement.textContent = `
				#root {
					filter: blur(${typeof blur === 'number' ? toPx(blur) : toPx(theme.modal.blur)})
				}
			`;
			document.head.append(styleElement);

			return () => {
				document.head.removeChild(styleElement);
			};
		}
	}, [blur, theme]);

	return createPortal(
		<Background
			ref={backgroundElement}
			data-background-color={theme.modal.backgroundColor}
			data-max-child-height={maxChildHeight}
			data-max-child-width={maxChildWidth}
			onClick={handleClick}
			{...props}
		>
			{children}
		</Background>,
		document.body
	);
};
