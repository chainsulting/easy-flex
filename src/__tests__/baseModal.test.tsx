import { render } from '@testing-library/react';
import 'jest-styled-components';
import React from 'react';
import { BaseModal } from '../baseModal/BaseModal';

describe('<BaseModal />', () => {
	test('should have the basic css settings', async () => {
		const result = render(<BaseModal onClose={() => undefined} />);
		expect(result.baseElement).toMatchSnapshot();
	});
});
