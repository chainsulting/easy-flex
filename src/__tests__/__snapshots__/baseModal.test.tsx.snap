// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseModal /> should have the basic css settings 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  box-sizing: border-box;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background-color: rgba(0,0,0,0.5);
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  margin: 0;
  padding: 0;
}

<body>
  <div />
  <div
    class="c0"
    data-background-color="rgba(0, 0, 0, 0.5)"
  />
</body>
`;
